lib/src/rules/void_checks.dart

```diff
@@ -131,10 +133,12 @@ class _Visitor extends SimpleAstVisitor<void> {
     checkedNode ??= node;
     if (expectedType == null || type == null) {
       return;
-    } else if (expectedType.isVoid &&
-            !isTypeAcceptableWhenExpectingVoid(type) ||
-        expectedType.isDartAsyncFutureOr &&
-            (expectedType as InterfaceType).typeArguments.first.isVoid &&
+    }
+
+    var futureOrVoidType =
+        context.typeProvider.futureOrType2(context.typeProvider.voidType);
+    if (expectedType.isVoid && !isTypeAcceptableWhenExpectingVoid(type) ||
+        expectedType == futureOrVoidType &&
```

Comment:
It seems to me that we are doing it the opposite way here - we should not construct `FutureOr<void>` and then use `==` so do force the analyzer to deconstruct it again. Instead check `isDartAsyncFutureOr` and that the type argument `isVoid`.

Reply:
I also looked at `isTypeAcceptableWhenExpectingVoid` and it seems that it can be improved by recursion instead of repeating the checks for `void` and `Null`.

Reply:
Ah I understand. OK fixed both issues.